components/webext-storage/src/sync/outgoing.rs

```diff
@@ -0,0 +1,185 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// The "outgoing" part of syncing - building the payloads to upload and
+// managing the sync state of the local DB.
+
+use interrupt_support::Interruptee;
+use rusqlite::{Row, Transaction};
+use sql_support::ConnExt;
+use sync15_traits::ServerTimestamp;
+use sync_guid::Guid as SyncGuid;
+
+use crate::error::*;
+
+use super::ServerPayload;
+
+fn outgoing_from_row(row: &Row<'_>) -> Result<(ServerPayload, i32)> {
+    let mirror_guid = row.get::<_, Option<SyncGuid>>("mirror_guid")?;
+    let staging_guid = row.get::<_, Option<SyncGuid>>("staging_guid")?;
+    // We could assert they are identical if non-none?
+    let guid = mirror_guid
+        .or(staging_guid)
+        .unwrap_or_else(SyncGuid::random);
+    let ext_id: String = row.get("ext_id")?;
+    let raw_data: Option<String> = row.get("data")?;
+    let (data, deleted) = if raw_data.is_some() {
+        (raw_data, false)
+    } else {
+        (None, true)
+    };
+    let change_counter = row.get::<_, i32>("sync_change_counter")?;
+    Ok((
+        ServerPayload {
+            ext_id,
+            guid,
+            data,
+            deleted,
+            last_modified: ServerTimestamp(0),
+        },
+        change_counter,
+    ))
+}
+
+/// Gets info about what should be uploaded and also records metadata about these
+/// items in a temp table. Returns a vec of the payloads which
+/// should be uploaded. record_uploaded() can be called after the upload is
+/// complete and the data in the temp table will be used to update the local
+/// store.
+pub fn get_and_record_outgoing(
+    tx: &Transaction<'_>,
+    _signal: &dyn Interruptee,
+) -> Result<Vec<ServerPayload>> {
+    // The item may not yet have a GUID (ie, it might not already be in either
+    // the mirror nor the incoming staging table.) If not for this fact, we
+    // could probably have arranged for the populating of the outgoing staging
+    // table to be done via a single `INSERT INTO ... SELECT` statement - but
+    // due to the GUID complication, and the fact we need to extract each
+    // record from this query anyway means we just do a single query, then
+    // perform the inserts one row at a time.
```

Comment:
I _think_ you can still do it like...

```sql
INSERT INTO storage_sync_outgoing_staging
(guid, ext_id, data, sync_change_counter)
SELECT coalesce(m.guid, s.guid, generate_guid()),
       l.ext_id, l.data, l.sync_change_counter
FROM storage_sync_data l
LEFT JOIN storage_sync_mirror m ON m.ext_id = l.ext_id
LEFT JOIN storage_sync_staging s ON s.ext_id = l.ext_id
WHERE sync_change_counter > 0
```

...And then read everything out of the temp table, but you're right—since we still need to return a `Vec<...>` of them, anyway, we may as well do it in Rust, and read it from the durable tables before staging! 👍